test/sqllogictest/column_knowledge.slt

```diff
@@ -0,0 +1,688 @@
+| Map (err: more than one record produced in subquery)
+| Project (#1)
+
+%3 =
+| Union %1 %2
+
+%4 =
+| Join %0 %3
+| | implementation = Differential %3 %0.()
+| | demand = (#0, #1)
+| Project (#0, #1)
+
+EOF
+
+query T multiline
+EXPLAIN SELECT * FROM t1 WHERE t1.f1 = 123 AND t1.f1 = (SELECT t2.f1 FROM t2);
+----
+%0 =
+| Get materialize.public.t1 (u1)
+| Filter (#0 = 123)
+| ArrangeBy ()
+
+%1 =
+| Get materialize.public.t2 (u3)
+| Filter (#0 = 123)
+| Project (#0)
+
+%2 =
+| Get materialize.public.t2 (u3)
+| Reduce group=()
+| | agg count(true)
+| Filter (err: more than one record produced in subquery), (#0 > 1)
+| Map (err: more than one record produced in subquery)
+| Project (#1)
+
+%3 =
+| Union %1 %2
+
+%4 =
+| Join %0 %3
+| | implementation = Differential %3 %0.()
+| | demand = (#0, #1)
+| Project (#0, #1)
+
+EOF
+
+#
+# Multipart keys
+#
+
+
+statement ok
+CREATE TABLE t4 (f1 INTEGER, f2 INTEGER, PRIMARY KEY (f1, f2));
+
+query T multiline
+EXPLAIN SELECT * FROM t4 AS a1, t4 AS a2 WHERE a1.f1 = 123 AND a1.f2 = 234 AND a1.f1 = a2.f1 AND a1.f2 = a2.f2;
+----
+%0 =
+| Get materialize.public.t4 (u7)
+| Filter (#0 = 123), (#1 = 234)
+| ArrangeBy ()
+
+%1 =
+| Get materialize.public.t4 (u7)
+| Filter (#0 = 123), (#1 = 234)
+
+%2 =
+| Join %0 %1
+| | implementation = Differential %1 %0.()
+| | demand = (#0..#3)
+
+EOF
+
+query T multiline
+EXPLAIN SELECT * FROM t4 AS a1 LEFT JOIN t4 AS a2 USING (f1, f2) WHERE a1.f1 = 123 AND a1.f2 = 234;
+----
+%0 = Let l0 =
+| Get materialize.public.t4 (u7)
+| Filter (#0 = 123), (#1 = 234)
+
+%1 =
+| Get %0 (l0)
+| ArrangeBy ()
+
+%2 =
+| Get materialize.public.t4 (u7)
+| Filter (#0 = 123), (#1 = 234)
+
+%3 = Let l1 =
+| Join %1 %2
+| | implementation = Differential %2 %1.()
+| | demand = (#0, #1)
+
+%4 =
+| Get %3 (l1)
+| Distinct group=(123, 234)
+| Negate
+
+%5 =
+| Get %0 (l0)
+| Distinct group=(123, 234)
+
+%6 =
+| Union %4 %5
+| ArrangeBy ()
+
+%7 =
+| Join %6 %0
+| | implementation = Differential %0 %6.()
+| | demand = (#0, #1)
+| Map null, null
+| Project (#0, #1, #4, #5)
+
+%8 =
+| Union %3 %7
+| Project (#0, #1)
+
+EOF
+
+#
+# Propagation in opposite direction
+#
+
+query T multiline
+EXPLAIN SELECT * FROM t4 AS a1 LEFT JOIN t4 AS a2 USING (f1, f2) WHERE a1.f1 = 123 AND a2.f2 = 234;
+----
+%0 = Let l0 =
+| Get materialize.public.t4 (u7)
+| Filter (#0 = 123), (#1 = 234)
+
+%1 =
+| Get %0 (l0)
+| ArrangeBy ()
+
+%2 =
+| Get materialize.public.t4 (u7)
+| Filter (#0 = 123), (#1 = 234)
+
+%3 = Let l1 =
+| Join %1 %2
+| | implementation = Differential %2 %1.()
+| | demand = (#0, #1)
+
+%4 =
+| Get %3 (l1)
+| Distinct group=(123, 234)
+| Negate
+
+%5 =
+| Get %0 (l0)
+| Distinct group=(123, 234)
+
+%6 =
+| Union %4 %5
+| ArrangeBy ()
+
+%7 =
+| Join %6 %0
+| | implementation = Differential %0 %6.()
+| | demand = (#0, #1)
+| Map null, null
+| Project (#0, #1, #4, #5)
+
+%8 =
+| Union %3 %7
+| Project (#0, #1)
+
+EOF
+
+#
+# Impossible conditions after propagation are currently not detected as such
```

Comment:
We are not detecting all sorts of impossible conditions, so hopefully in the future this plan will flip.